extension/parquet/parquet_extension.cpp

```diff
@@ -81,17 +88,38 @@ struct ParquetReadLocalState : public LocalTableFunctionState {
 
 enum class ParquetFileState : uint8_t { UNOPENED, OPENING, OPEN, CLOSED };
 
+struct ParquetFileReaderData {
+	// Create data for an unopened file
+	ParquetFileReaderData(const string &file_to_be_opened)
+	    : reader(nullptr), file_state(ParquetFileState::UNOPENED), file_mutex(make_uniq<mutex>()),
+	      file_to_be_opened(file_to_be_opened) {
+	}
+	// Create data for an existing reader
+	ParquetFileReaderData(shared_ptr<ParquetReader> reader_p)
+	    : reader(std::move(reader_p)), file_state(ParquetFileState::OPEN), file_mutex(make_uniq<mutex>()) {
+	}
+
+	//! Currently opened reader for the file
+	shared_ptr<ParquetReader> reader;
+	//! Flag to indicate the file is being opened
+	ParquetFileState file_state;
+	//! Mutexes to wait for the file when it is being opened
+	unique_ptr<mutex> file_mutex;
```

Comment:
Can we turn this into a regular mutex?

Reply:
vector doesn't like that: mutex is not movable. 

The idea is that with this you can:
- get a ref to this mutex
- drop the global state lock
- grab the file lock
- release the file lock
- regrab the global state lock

with the idea of opening a new file without locking global state while doing so: in the meantime the vector holding the mutex may get resized.